modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,142 +1,272 @@
-import * as Fun from './Fun';
+import * as Type from './Type';
 
-/** A value which may be Optional.some(a) or Optional.none().
- *  Useful for representing partiality or pass/fail conditions that produce a value when successful.
- *  Sometimes called a "type-safe alternative to null".
- *  Can be thought of as a list of exactly zero or 1 elements.
- */
-export interface Optional<T> {
-  /** If none, run whenNone; if some(a) run whenSome(a) */
-  readonly fold: <T2> (whenNone: () => T2, whenSome: (v: T) => T2) => T2;
+/**
+ The `Optional` type represents a value (of any type) that potentially does not
+ exist. Any `Optional<T>` can either be a `Some<T>` (in which case the value
+ does exist) or a `None` (in which case the value does not exist). This file
+ defines a whole lot of FP-inspired utility functions for dealing with
+ `Optional` objects. Note: Functions are grouped based on, primarily, their
+ Haskell roots. That said, I'm not a Haskell programmer so it might be
+ imperfect.
 
-  readonly isSome: () => boolean;
-  readonly isNone: () => boolean;
+ Comparison with null or undefined
+ - We don't get fancy null coalescing operators with `Optional`
+ - We do get fancy helper functions with `Optional`
+ - `Optional` support nesting, and allow for the type to still be nullable (or
+ another `Optional`)
+ - There is no option to turn off strict-optional-checks like there is for
+ strict-null-checks
+ - Try to use `Optional` instead of null or undefined where you can
+*/
```

Comment:
```suggestion
 Comparison with `null` or `undefined`
 - We don't get fancy null coalescing operators with `Optional` (e.g. `obj?.val`)
 - We do get fancy helper functions with `Optional`
 - `Optional` supports nesting, and allow for the type to still be nullable (or
 another `Optional`)
 - There is no option to turn off strict-optional-checks like there is for
 strict-null-checks
 - Try to use `Optional` instead of `null` or `undefined` where you can
*/
```